src/Symfony/Component/HttpClient/CachingHttpClient.php

```diff
@@ -11,68 +11,365 @@
 
-        $kernel = new HttpClientKernel($client);
-        $this->cache = new HttpCache($kernel, $store, null, $defaultOptions);
+            if (!class_exists(HttpClientKernel::class)) {
+                throw new \LogicException(\sprintf('Using "%s" requires the HttpKernel component, try running "composer require symfony/http-kernel".', __CLASS__));
+            }
+
+            $kernel = new HttpClientKernel($client);
+            $this->cache = new HttpCache($kernel, $cache, null, $defaultOptions);
 
-        unset($defaultOptions['debug']);
-        unset($defaultOptions['default_ttl']);
-        unset($defaultOptions['private_headers']);
-        unset($defaultOptions['skip_response_headers']);
-        unset($defaultOptions['allow_reload']);
-        unset($defaultOptions['allow_revalidate']);
-        unset($defaultOptions['stale_while_revalidate']);
-        unset($defaultOptions['stale_if_error']);
-        unset($defaultOptions['trace_level']);
-        unset($defaultOptions['trace_header']);
+            unset($defaultOptions['debug']);
+            unset($defaultOptions['default_ttl']);
+            unset($defaultOptions['private_headers']);
+            unset($defaultOptions['skip_response_headers']);
+            unset($defaultOptions['allow_reload']);
+            unset($defaultOptions['allow_revalidate']);
+            unset($defaultOptions['stale_while_revalidate']);
+            unset($defaultOptions['stale_if_error']);
+            unset($defaultOptions['trace_level']);
+            unset($defaultOptions['trace_header']);
+        } else {
+            $this->cache = $cache;
+        }
 
         if ($defaultOptions) {
             [, $this->defaultOptions] = self::prepareRequest(null, null, $defaultOptions, $this->defaultOptions);
         }
     }
 
     public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        if ($this->cache instanceof HttpCache) {
+            return $this->legacyRequest($method, $url, $options);
+        }
+
+        [$fullUrl, $options] = self::prepareRequest($method, $url, $options, $this->defaultOptions);
+
+        $fullUrl = implode('', $fullUrl);
+        $fullUrlTag = self::hash($fullUrl);
+
+        if (\in_array($method, self::UNSAFE_METHODS, true)) {
+            $this->cache->invalidateTags([$fullUrlTag]);
+        }
+
+        if ('' !== $options['body'] || isset($options['normalized_headers']['range']) || !\in_array($method, self::CACHEABLE_METHODS, true)) {
+            return new AsyncResponse($this->client, $method, $url, $options);
+        }
+
+        $requestHash = self::hash($method.$fullUrl);
+        $varyKey = "vary_{$requestHash}";
+        $varyFields = $this->cache->get($varyKey, static fn (): array => []);
+
+        $metadataKey = self::getMetadataKey($requestHash, $options['normalized_headers'], $varyFields);
+        $cachedData = $this->cache->get($metadataKey, static fn (): null => null);
```

Comment:
do we want to save the null value in the backed on a miss?
if not, we should use the $save argument passed by ref to the closure:
```suggestion
        $cachedData = $this->cache->get($metadataKey, fn ($item, &$save) => $save = false;)
```

Reply:
(the same concern should be consider everywhere get() is used)